File: clojure/anagram/anagram.clj
(ns anagram
  (:require [clojure.string :as string]))

(defn- normalize [text]
  (string/lower-case text))

(defn- anagrams? [a b]
  (let [c (normalize a)
        d (normalize b)]
    (and (not= c d)
         (= (sort c) (sort d)))))

(defn anagrams-for [text candidates]
  (filter (partial anagrams? text) candidates))

File: javascript/anagram/anagram.js
var Anagram = function(subject) {
  this.subject = subject;
};

Anagram.prototype.matches = function(candidates) {
  if (arguments.length > 1) { // support varargs
    candidates = Array.prototype.slice.call(arguments);
  }
  return findAnagrams(this.subject, toArray(candidates));
};

function findAnagrams(subject, candidates) {
  return candidates.filter(function(word) { return isAnagram(subject, word); });
}

function isAnagram(a, b) {
  a = a.toLowerCase();
  b = b.toLowerCase();
  return a !== b && sortWord(a) === sortWord(b);
}

function sortWord(word) {
  return word.split("").sort().join("");
}

function toArray(object) {
  return [].concat(object);
}

module.exports = Anagram;
